python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -2729,7 +2749,10 @@ def transfer_from(
                 "" if len(direct_transfers) == 1 else "s",
             )
 
-        self._register_datasets(artifacts)
+        # We are overwriting previous datasets that may have already
+        # existed. We therefore should ensure that we force the
+        # datastore records to agree.
```

Comment:
Could it happen that old and new datasets have different set of components?

Reply:
I've added an extra commentary but still use replace. We really need to sort this out properly.